framework/src/utils/MooseMeshXYCuttingUtils.C

```diff
@@ -0,0 +1,1230 @@
+                       const std::vector<boundary_id_type> boundary_ids_for_side_0,
+                       const std::vector<boundary_id_type> boundary_ids_for_side_2)
+{
+  BoundaryInfo & boundary_info = mesh.get_boundary_info();
+  Elem * elem_Tri3_new = mesh.add_elem(new Tri3);
+  elem_Tri3_new->set_node(0) = mesh.node_ptr(node_id_0);
+  elem_Tri3_new->set_node(1) = mesh.node_ptr(node_id_1);
+  elem_Tri3_new->set_node(2) = mesh.node_ptr(node_id_2);
+  for (const auto & boundary_id_for_side_0 : boundary_ids_for_side_0)
+    boundary_info.add_side(elem_Tri3_new, 0, boundary_id_for_side_0);
+  for (const auto & boundary_id_for_side_1 : boundary_ids_for_side_1)
+    boundary_info.add_side(elem_Tri3_new, 1, boundary_id_for_side_1);
+  for (const auto & boundary_id_for_side_2 : boundary_ids_for_side_2)
+    boundary_info.add_side(elem_Tri3_new, 2, boundary_id_for_side_2);
+  elem_Tri3_new->subdomain_id() = subdomain_id;
+}
+
+void
+makeOrderedNodeList(std::vector<std::pair<dof_id_type, dof_id_type>> & node_assm,
+                    std::vector<dof_id_type> & ordered_node_list,
+                    std::vector<dof_id_type> & elem_id_list,
+                    std::vector<dof_id_type> & ordered_elem_id_list)
+{
+  // a flag to indicate if the ordered_node_list has been reversed
+  bool isFlipped = false;
+  // Start from the first element, try to find a chain of nodes
+  ordered_node_list.push_back(node_assm.front().first);
+  ordered_node_list.push_back(node_assm.front().second);
+  ordered_elem_id_list.push_back(elem_id_list.front());
+  // Remove the element that has been added to ordered_node_list
+  node_assm.erase(node_assm.begin());
+  elem_id_list.erase(elem_id_list.begin());
+  const unsigned int node_assm_size_0 = node_assm.size();
+  for (unsigned int i = 0; i < node_assm_size_0; i++)
+  {
+    // Find nodes to expand the chain
+    dof_id_type end_node_id = ordered_node_list.back();
+    auto isMatch1 = [end_node_id](std::pair<dof_id_type, dof_id_type> old_id_pair)
+    { return old_id_pair.first == end_node_id; };
+    auto isMatch2 = [end_node_id](std::pair<dof_id_type, dof_id_type> old_id_pair)
+    { return old_id_pair.second == end_node_id; };
+    auto result = std::find_if(node_assm.begin(), node_assm.end(), isMatch1);
+    bool match_first;
+    if (result == node_assm.end())
+    {
+      match_first = false;
+      result = std::find_if(node_assm.begin(), node_assm.end(), isMatch2);
+    }
+    else
+    {
+      match_first = true;
+    }
+    // If found, add the node to boundary_ordered_node_list
+    if (result != node_assm.end())
+    {
+      ordered_node_list.push_back(match_first ? (*result).second : (*result).first);
+      node_assm.erase(result);
+      const auto elem_index = std::distance(node_assm.begin(), result);
+      ordered_elem_id_list.push_back(elem_id_list[elem_index]);
+      elem_id_list.erase(elem_id_list.begin() + elem_index);
```

Comment:
any reason you have these erasing?
erasing on a vector like this will trigger a re-allocation every time

Reply:
I think for node_assm you do want to erase so that the node isnt found twice

Reply:
Here I am trying to find the elem ids corresponding to the node pairs so they are erased accordingly.